float32 x
float32 y
float32 z
float32 radius
float32 small_radius
int32 type
